### DISCUSSION TITLE
Does meson support aarch64-elf-gcc?

### DISCUSSION THREAD

**Discussion by noaxie** [mesonbuild/meson#11384] [Q&A] [answered]
I would like to setup a cross project using the linker aarch64-elf-gcc but it doesn't seem to be supported. (At least in the linkers mentions [here](https://mesonbuild.com/Reference-tables.html#linker-ids)) Is there any way i can force meson to use my desired linker?

Would setting c_ld inside the cross file force meson to use aarch64-elf-gcc as the linker?

**Comment by eli-schwartz** [comment, marked as answer]
Is GCC the linker or is it the linker *driver* and the actual linker is binutils bfd or some such?

**Reply by noaxie** [reply]
I believe it's the linker driver. (Is there anyway for me to confirm this?)

**Reply by eli-schwartz** [reply]
So then it should just work fine by default. Meson will run this command to check what the linker ID is:
```
aarch64-elf-gcc -Wl,--version
```

It will then compare the output and try to extract linker names (such as detecting "GNU gold" or "GNU ld" in the ouput) to identify which linker is going to be internally used by the compiler's "linker driver" mode.

This may be influenced by the use of `-fuse-ld=XXX` either in $CFLAGS or when set as the `c_ld` option in the machine file (c_ld, when set, will tell Meson to add `-fuse-ld=XXX` with the value you specified).

...

Of course things are a bit different for Windows, where the linker is often invoked directly. Meson has handling for that as well, the gory details are in `mesonbuild/compilers/detect.py` and `mesonbuild/linkers/detect.py`, but I think that focusing on GCC is probably fine for this discussion.

**Reply by noaxie** [reply]
Ok understood thanks for the good explanation. Sorry for my dumb question but I would like to further my understanding of meson and I have one little follow up question. What does meson do when the compiler doesn't have the fuse-ld option such as the compcert compiler? 

So for example if I wanted to compile with the compcert compiler but use a custom linker such as aarch64-elf-gcc?

**Reply by eli-schwartz** [reply]
We just hardcode `linkers.CompCertDynamicLinker` as the linker class (so the ID is `'ccomp'`), and only implement `compiler.use_linker_args(cls, linker, comp_version)` for GNU-like and clang-like compilers.

There is currently no support for detecting two different compilers for the same language and using one to compile but the other to link. You can, however, detect both a C and a C++ compiler and specify which language to link a target as regardless of the individual source files. In theory, you might be able to then compile C with compcert but link it with a `g++`?

I'm not sure the topic has come up before. If there's a compelling reason to do this then you could submit a dedicated issue requesting such a thing be added, I guess.
